larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,623 @@
+        }
+        catch (...) {
+        CaloHitList hits;
+        cluster->GetOrderedCaloHitList().FillCaloHitList(hits);
+        centroid11 = hits.front()->GetPositionVector();
+        centroid12 = hits.front()->GetPositionVector();
+        }
+
+        const CartesianVector centroid1((centroid11 + centroid12) * 0.5f);
+	
+	this->EdgeHitFinder(cluster, clusterEdgeHits); 
+
+        const unsigned int clusterNHits{cluster->GetNCaloHits()};
+        const unsigned long int clusterNEdgeHits{clusterEdgeHits.size()};
+
+	for (auto iter2 = std::next(iter) ; iter2 != pClusterList->end() ; ++iter2)
+	{
+	    const Cluster *const otherCluster(*iter2);
+	    const MCParticle *otherClusterMC(this->GetMCForCluster(otherCluster, clusterToMCParticleMap));
+            sigBac = 0;
+
+            if (clusterMC == otherClusterMC)
+            {
+                   sigBac = 1;
+            }
+
+	    CaloHitList otherClusterEdgeHits;
+	    CartesianVector centroid21(0.f, 0.f, 0.f);
+	    CartesianVector centroid22(0.f, 0.f, 0.f);
+	    try {
+	    centroid21 = otherCluster->GetCentroid(otherCluster->GetInnerPseudoLayer());
+            centroid22 = otherCluster->GetCentroid(otherCluster->GetOuterPseudoLayer());
+	    }
+	    catch (...) {
+	    CaloHitList hits;
+            cluster->GetOrderedCaloHitList().FillCaloHitList(hits);
+            centroid21 = hits.front()->GetPositionVector();
+            centroid22 = hits.front()->GetPositionVector();
+	    }
+	    const CartesianVector centroid2((centroid21 + centroid22) * 0.5f);
+	    const float centroidSeparation{std::sqrt(centroid1.GetDistanceSquared(centroid2))};
+	    const CartesianVector centroidVector(centroid2.GetX() - centroid1.GetX(), 0.f, centroid2.GetZ() - centroid1.GetZ());
+	    
+	    const VertexList *pVertexList{nullptr};
+            float vtxClusterAngle{-1};
+	    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_INITIALIZED, !=, PandoraContentApi::GetList(*this,                       m_vertexListName, pVertexList));
+	    
+	    if (pVertexList && !pVertexList->empty())
+            {
+                const Vertex* pVertex{nullptr}; 
+   	        pVertex = pVertexList->front();
+	        const HitType hitType(LArClusterHelper::GetClusterHitType(cluster));
+                CartesianVector vertexPosition(LArGeometryHelper::ProjectPosition(this->GetPandora(), pVertex->GetPosition() , hitType));
+
+		const CartesianVector vtxToCluster(centroid1.GetX() - vertexPosition.GetX(), 0.f, centroid1.GetZ() - vertexPosition.GetZ());
+                const CartesianVector vtxToOtherCluster(centroid2.GetX() - vertexPosition.GetX(), 0.f, centroid2.GetZ() - vertexPosition.GetZ());
+		const float D1(vtxToCluster.GetMagnitude());
+		const float D2(vtxToOtherCluster.GetMagnitude());
+	        if ( D1*D2 != 0)
+		{
+		    vtxClusterAngle = (vtxToCluster.GetDotProduct(vtxToOtherCluster) / (D1 * D2));
+		}
+	    }
+
+	    this->EdgeHitFinder(otherCluster, otherClusterEdgeHits);
```

Comment:
Seems like this function is going to be called repeatedly (as the iterator loops through the clusters).  Would it be better to loop through the clusters once and fill a map?